modules/porous_flow/include/userobjects/PorousFlowDictator.h

```diff
@@ -0,0 +1,175 @@
+ * well as the number of fluid phases and
+ * the number of fluid components.
+ *
+ * All PorousFlow Materials and Kernels calculate
+ * and use derivatives with respect to all the variables
+ * mentioned in this Object, at least in principal
+ * (in practice they may be lazy and not compute
+ * all derivatives).
+ */
+class PorousFlowDictator :
+  public GeneralUserObject,
+  public Coupleable,
+  public ZeroInterface
+{
+ public:
+  PorousFlowDictator(const InputParameters & parameters);
+
+  void initialize();
+  void execute();
+  void finalize();
+
+  /**
+   * The number of PorousFlow variables.  Materials
+   * and Kernels will calculate and use derivatives
+   * with respect to these variables in the Jacobian
+   */
+  unsigned int num_variables() const;
+
+  /// the number of fluid phases
+  unsigned int num_phases() const;
+
+  /// the number of fluid components
+  unsigned int num_components() const;
+
+  /**
+   * the PorousFlow variable number
+   * @param moose_var_num the MOOSE variable number
+   * eg if porflow_vars = 'pwater pgas', and the variables in
+   * the simulation are 'energy pwater pgas shape'
+   * then porflow_var_num(2) = 1
+   */
+  unsigned int porflow_var_num(unsigned int moose_var_num) const;
+
+  /**
+   * returns true if moose_var_num is not a porflow var
+   * @param moose_var_num the MOOSE variable number
+   * eg if porflow_vars = 'pwater pgas', and the variables in
+   * the simulation are 'energy pwater pgas shape'
+   * then not_porflow_var(0) = true, not_porflow_var(1) = false
+   */
+  bool not_porflow_var(unsigned int moose_var_num) const;
+
+  /**
+   * a vector of pointers to VariableValues
+   * @param porflow_var_num the porflow variable number
+   * eg if porflow_vars = 'pwater pgas', then
+   * (*porflow_vals(1))[qp] = pgas evaluated at quadpoint qp
+   * Also porflow_vals(i) = &coupledValue
+   */
+  const VariableValue * porflow_vals(unsigned int porflow_var_num) const;
+
+  /**
+   * a vector of pointers to old VariableValues
+   * @param porflow_var_num the porflow variable number
+   * eg if porflow_vars = 'pwater pgas', then
+   * (*porflow_vals_old(1))[qp] = old pgas evaluated at quadpoint qp
+   * Also porflow_vals_old(i) = &coupledValueOld
+   */
+  const VariableValue * porflow_vals_old(unsigned int porflow_var_num) const;
+
+  /**
+   * a vector of pointers to grad(Variable)
+   * @param porflow_var_num the porflow variable number
+   * eg if porflow_vars = 'pwater pgas', then
+   * (*grad_var(1))[qp] = grad(pgas) evaluated at quadpoint qp
+   * Also grad_var(i) = &coupledGradient
+   */
+  const VariableGradient * grad_var(unsigned int porflow_var_num) const;
+
+  /**
+   * The nodal variable values for the given porflow_var_num
+   * To extract the value of porflow variable "pvar", at
+   * node i, use (*PorousFlowDictator.nodal_var(pvar))[i]
+   * @param porflow_var_num the porflow variable number
+   */
+  const VariableValue * nodal_var(unsigned int porflow_var_num) const;
```

Comment:
@WilkAndy, have you tested this with threading? I'm probably jumping the gun as I haven't made it all the way to the .C files yet... But GeneralUserObjects are not threaded, so all threads in threaded objects see the same UO. That may lead to VariableValues for mismatching threads to being used).